### PR TITLE
[BugFix] fix mv rewrite bug for join/agg with limit

### LINKED ISSUE #22679: mv rewrite failed for join with limit
### Steps to reproduce the behavior (Required)
1. CREATE TABLE
CREATE TABLE `aggregate_table_with_null` (
  `k1` date NULL COMMENT "",
  `k2` datetime NULL COMMENT "",
  `k3` char(20) NULL COMMENT "",
  `k4` varchar(20) NULL COMMENT "",
  `k5` boolean NULL COMMENT "",
  `v1` bigint(20) SUM NULL COMMENT "",
  `v2` bigint(20) SUM NULL COMMENT "",
  `v3` bigint(20) SUM NULL COMMENT "",
  `v4` bigint(20) MAX NULL COMMENT "",
  `v5` largeint(40) MAX NULL COMMENT "",
  `v6` float MIN NULL COMMENT "",
  `v7` double MIN NULL COMMENT "",
  `v8` decimal128(38, 9) SUM NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`, `k4`, `k5`) BUCKETS 3
PROPERTIES (
"replication_num" = "3",
"in_memory" = "false",
"storage_format" = "V2",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"compression" = "LZ4"
);

CREATE TABLE `duplicate_table_with_null_partition` (
  `k1` date NULL COMMENT "",
  `k2` datetime NULL COMMENT "",
  `k3` char(20) NULL COMMENT "",
  `k4` varchar(20) NULL COMMENT "",
  `k5` boolean NULL COMMENT "",
  `k6` tinyint(4) NULL COMMENT "",
  `k7` smallint(6) NULL COMMENT "",
  `k8` int(11) NULL COMMENT "",
  `k9` bigint(20) NULL COMMENT "",
  `k10` largeint(40) NULL COMMENT "",
  `k11` float NULL COMMENT "",
  `k12` double NULL COMMENT "",
  `k13` decimal128(27, 9) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`k1`, `k2`, `k3`, `k4`, `k5`)
COMMENT "OLAP"
PARTITION BY RANGE(`k1`)
(PARTITION p202006 VALUES [("0000-01-01"), ("2020-07-01")),
PARTITION p202007 VALUES [("2020-07-01"), ("2020-08-01")),
PARTITION p202008 VALUES [("2020-08-01"), ("2020-09-01")))
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`, `k4`, `k5`) BUCKETS 3
PROPERTIES (
"replication_num" = "3",
"in_memory" = "false",
"storage_format" = "V2",
"enable_persistent_index" = "false",
"replicated_storage" = "true",
"compression" = "LZ4"
);

CREATE MATERIALIZED VIEW `mv19`
COMMENT "MATERIALIZED_VIEW"
DISTRIBUTED BY HASH(`k1`) BUCKETS 10
REFRESH ASYNC
PROPERTIES (
"replication_num" = "3",
"replicated_storage" = "true",
"storage_medium" = "HDD"
)
AS SELECT `t1`.`k1`, `t2`.`k2`
FROM `test_mv_async_db_22c7b56e_e3d8_11ed_a37d_00163e0e489a`.`duplicate_table_with_null_partition` AS `t1` INNER JOIN `test_mv_async_db_22c7b56e_e3d8_11ed_a37d_00163e0e489a`.`aggregate_table_with_null` AS `t2` ON `t1`.`k1` = `t2`.`k1`;

3. INSERT INTO data
4. query
explain SELECT t1.k1 k1, t2.k2 k2 from duplicate_table_with_null_partition t1 join aggregate_table_with_null t2 on t1.k1 = t2.k1 limit 10

### Expected behavior (Required)
the query is rewritten by mv19
### Real behavior (Required)
the query is not rewritten by mv10
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #22679 

## Problem Summary(Required):

fix mv rewrite bug for join/agg with limit. fix it by adding limit into rewritten root operator if the query has limit.
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|6|6|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 1]
```diff
@@ -129,6 +130,15 @@ public CostEstimate visitOperator(Operator node, ExpressionContext context) {
         public CostEstimate visitPhysicalOlapScan(PhysicalOlapScanOperator node, ExpressionContext context) {
             Statistics statistics = context.getStatistics();
             Preconditions.checkNotNull(statistics);
+            if (node.getTable().isMaterializedView()) {
+                ColumnRefSet usedColumns = statistics.getUsedColumns();
+                Projection projection = node.getProjection();
+                if (projection != null) {
+                    // remove projection keys
+                    usedColumns.except(projection.getColumnRefMap().keySet());
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:138]: Why exception projection? will add its costs later?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 2]
```diff
@@ -129,6 +130,15 @@ public CostEstimate visitOperator(Operator node, ExpressionContext context) {
...
+                }
+                return CostEstimate.of(statistics.getOutputSize(usedColumns), 0, 0);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:140]: used column contains predicate columns? if not, why not compute this cost?

Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:140]: yes, used column contains predicate columns.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 3]
```diff
@@ -347,6 +347,11 @@ private OptExpression tryRewrite(List<Table> queryTables,
 
             OptExpression rewrittenExpression = tryRewriteForRelationMapping(rewriteContext, compensationJoinColumns);
             if (rewrittenExpression != null) {
+                // copy limit into rewritten plan
+                // limit will affect the statistics of rewritten plan
+                if (rewriteContext.getQueryExpression().getOp().hasLimit()) {
+                    rewrittenExpression.getOp().setLimit(rewriteContext.getQueryExpression().getOp().getLimit());
+                }
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:354]: Now, MV Rewrite and limit push-down rules are already separated. mv rewrite rules are applied after limit pushdown rules.
if limit is pushed down, the upper node will keep the limit, we should also add limit back into the rewritten plan.

**Review by Youngwb — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Review by LiShuMing — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**